GAE/src/org/akvo/flow/rest/FormAssemblyRestService.java

```diff
@@ -0,0 +1,411 @@
+                PublishedForm.generate(jacksonForm));
+    }
+
+    private Survey assembleForm(SurveyDto surveyDto) {
+        Survey form = getFormFromDto(surveyDto);
+        attachFormTranslations(form);
+        List<Question> questions = getQuestionList(form.getQuestionGroupMap());
+        attachCascadeResources(questions);
+        return form;
+    }
+
+    private void formUploadSuccess(SurveyGroup survey, Survey form) {
+        List<Question> questions = getQuestionList(form.getQuestionGroupMap());
+        form.setStatus(Survey.Status.PUBLISHED);
+        if (form.getWebForm()) {
+            boolean webForm = WebForm.validWebForm(survey, form, questions);
+            form.setWebForm(webForm);
+        }
+        new SurveyDAO().save(form);
+        //invalidate any cached reports in flow-services
+        //TODO: shall this be done in another service?
+        List<Long> ids = new ArrayList<>();
+        ids.add(form.getObjectId());
+        SurveyUtils.notifyReportService(ids, "invalidate");
+    }
+
+    private Survey getFormFromDto(SurveyDto surveyDto) {
+        Survey form = new Survey();
+        form.setKey(KeyFactory.createKey("Survey", surveyDto.getKeyId()));
+        form.setCode(surveyDto.getCode());
+        form.setName(surveyDto.getName());
+        form.setVersion(Double.parseDouble(surveyDto.getVersion()));
+        form.setDesc(surveyDto.getDescription());
+        form.setStatus(Survey.Status.NOT_PUBLISHED);
+        form.setPath(surveyDto.getPath());
+        form.setSurveyGroupId(surveyDto.getSurveyGroupId());
+        form.setDefaultLanguageCode(surveyDto.getDefaultLanguageCode());
+        form.setRequireApproval(surveyDto.getRequireApproval());
+        form.setCreatedDateTime(surveyDto.getCreatedDateTime());
+        form.setLastUpdateDateTime(surveyDto.getLastUpdateDateTime());
+        form.setAncestorIds(surveyDto.getAncestorIds());
+        TreeMap<Integer, QuestionGroup> groupMap = getGroupMap(surveyDto);
+        form.setQuestionGroupMap(groupMap);
+        return form;
+    }
+
+    @Nonnull
+    private TreeMap<Integer, QuestionGroup> getGroupMap(SurveyDto surveyDto) {
+        TreeMap<Integer, QuestionGroup> groupMap = new TreeMap<>();
+        List<QuestionGroupDto> groupDtos = surveyDto.getQuestionGroupList();
+        if (groupDtos != null) {
+            int i = 1;
+            for (QuestionGroupDto groupDto : groupDtos) {
+                QuestionGroup group = mapToGroup(groupDto);
+                if (groupMap.containsKey(groupDto.getOrder())) {
+                    groupMap.put(i, group);
+                    groupDto.setOrder(i);
+                } else {
+                    int order = groupDto.getOrder() != null ? groupDto.getOrder() : i;
+                    groupMap.put(order, group);
+                }
+                i++;
+            }
+        }
+        return groupMap;
+    }
+
+    private QuestionGroup mapToGroup(QuestionGroupDto groupDto) {
```

Comment:
Maybe you want to group this and other mapping methods in the respective mapper? i.e. `QuestionDtoMapper`, `QuestionGroupDtomapper`, `QuestionOptionDtoMapper`

Reply:
My idea was to do that when adding the tests, extract those into different delegate classes to be able to easily test them.